enzyme/Enzyme/MLIR/Analysis/ActivityAnnotations.cpp

```diff
@@ -0,0 +1,1319 @@
+  if (map.empty()) {
+    os << "<empty>\n";
+    return;
+  }
+  for (const auto &[aliasClass, origins] : map) {
+    os << "  " << aliasClass << " originates from {";
+    if (origins.isUnknown()) {
+      os << "<unknown>";
+    } else if (origins.isUndefined()) {
+      os << "<undefined>";
+    } else {
+      llvm::interleaveComma(origins.getElements(), os);
+    }
+    os << "}\n";
+  }
+}
+
+void enzyme::ForwardOriginsMap::print(raw_ostream &os) const {
+  printMapOfSetsLattice(this->map, os);
+}
+
+void enzyme::BackwardOriginsMap::print(raw_ostream &os) const {
+  printMapOfSetsLattice(this->map, os);
+}
+
+// bool enzyme::sortAttributes(Attribute a, Attribute b) {
+//   auto originA = dyn_cast<enzyme::ArgumentOriginAttr>(a);
+//   auto originB = dyn_cast<enzyme::ArgumentOriginAttr>(b);
+//   if (originA && originB)
+//     return originA.getArgNumber() < originB.getArgNumber();
+
+//   auto distinctA = dyn_cast<DistinctAttr>(a);
+//   auto distinctB = dyn_cast<DistinctAttr>(b);
+//   // If not distinct attributes, sort them arbitrarily.
+//   if (!(distinctA && distinctB))
+//     return &a < &b;
+
+//   auto pseudoA = dyn_cast_if_present<enzyme::PseudoAliasClassAttr>(
+//       distinctA.getReferencedAttr());
+//   auto pseudoB = dyn_cast_if_present<enzyme::PseudoAliasClassAttr>(
+//       distinctB.getReferencedAttr());
+//   auto strA = dyn_cast_if_present<StringAttr>(distinctA.getReferencedAttr());
+//   auto strB = dyn_cast_if_present<StringAttr>(distinctB.getReferencedAttr());
+//   if (pseudoA && pseudoB) {
+//     return std::make_pair(pseudoA.getArgNumber(), pseudoA.getDepth()) <
+//            std::make_pair(pseudoB.getArgNumber(), pseudoB.getDepth());
+//   } else if (strA && strB) {
+//     return strA.strref() < strB.strref();
+//   }
+//   // Order pseudo/origin classes before fresh classes
+//   return (pseudoA || originA) && !(pseudoB || originB);
+// }
+
+void enzyme::DenseActivityAnnotationAnalysis::setToEntryState(
+    ForwardOriginsMap *lattice) {
+  auto *block = dyn_cast<Block *>(lattice->getAnchor());
+  if (!block)
+    return;
+
+  auto funcOp = cast<FunctionOpInterface>(block->getParentOp());
+  ChangeResult changed = ChangeResult::NoChange;
+  for (BlockArgument arg : funcOp.getArguments()) {
+    auto *argClass = getOrCreateFor<AliasClassLattice>(block, arg);
+    auto origin = ArgumentOriginAttr::get(FlatSymbolRefAttr::get(funcOp),
+                                          arg.getArgNumber());
+    changed |= lattice->insert(argClass->getAliasClassesObject(),
+                               ValueOriginSet(origin));
+  }
+  propagateIfChanged(lattice, changed);
+}
+
+std::optional<Value> getStored(Operation *op);
+std::optional<Value> getCopySource(Operation *op);
+
+LogicalResult enzyme::DenseActivityAnnotationAnalysis::visitOperation(
+    Operation *op, const ForwardOriginsMap &before, ForwardOriginsMap *after) {
+  join(after, before);
```

Comment:
Comment, no need for action: this is one of the reasons why the analysis may be inefficient. Specifically, this join subsumes `propagateIfChanged`, and we will call it again below. This may schedule two updates when one would have been sufficient.